### PR TITLE
Push down HTML Printer and various associated classes from o.e.jface …

### PR DESCRIPTION
**Pull request by robstryker**
…to org.eclipse.text

Signed-off-by: Rob Stryker <stryker@redhat.com>

Part 2

### CODE REVIEW

**Comment by robstryker** [comment]
This PR looks ready for review? I can only see 1 test fail on mac and it seems unrelated to my code?

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+ * object has been inserted.
+ * </p>
+ * <p>
+ * This class is a clone of org.eclipse.swt.RGB without any UI dependencies
+ * </p>
+ *
+ * @since 3.14
+ */
+
+public final class CoreRGB implements Serializable {
```
Comment (mickaelistria) [inline bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java]: As it's a new API, let's try to make it as clean as possible from the beginning:
* What about making it a record?
* What about calling it just RGB anyway (and people who need to disambiguize SWT vs text would just use package name)?

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+* @param hue the hue value for the HSB color (from 0 to 360)
+* @param saturation the saturation value for the HSB color (from 0 to 1)
+* @param brightness the brightness value for the HSB color (from 0 to 1)
+*
+* @exception IllegalArgumentException <ul>
+*    <li>ERROR_INVALID_ARGUMENT - if the hue is not between 0 and 360 or
+*    the saturation or brightness is not between 0 and 1</li>
+* </ul>
+*
+* @since 3.2
```
Comment (mickaelistria) [inline bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java]: All those seem incorrect as the class is new.

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+import org.eclipse.core.runtime.Platform;
+
+/**
+ * Provides a set of convenience methods for creating HTML pages.
+ * <p>
+ * Moved into this package from <code>org.eclipse.jface.internal.text.revisions</code>.</p>
+ *
+ * @since 3.14
+ */
+public class CoreHTMLPrinter {
```
Comment (mickaelistria) [inline bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java]: Can't we find a nicer name for this new API?

File: bundles/org.eclipse.jface.text/src/org/eclipse/jface/internal/text/html/SingleCharReader.java [hunk 4]
```diff
@@ -13,59 +13,15 @@
...
-	}
-
-	/**
-	 * @see Reader#ready()
-	 */
-	@Override
-	public boolean ready() throws IOException {
-		return true;
-	}
+public abstract class SingleCharReader extends SingleCharacterReader {
```
Comment (mickaelistria) [inline bundles/org.eclipse.jface.text/src/org/eclipse/jface/internal/text/html/SingleCharReader.java:26]: Let's add the official `@Deprecated` java annotation as well.
Should we plan and announce removal?

**Review by mickaelistria — APPROVED** [review]
The overall idea is more than sane.
However, as we're basically creating some new API, it seems best to use it as an opportunity to refresh the code rather than duplicating old stuff.

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java (hunk 1 above)
Comment (robstryker) [reply bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java]: I have no problem naming it RGB and using package to disambiguate. I think making it a record has a few drawbacks.  First, there's absolutely no reason to mix things up if we don't need to. I prefer this class match the org.eclipse.swt version as closely as possible, with the same exceptions being thrown, the same logic, and essentially the same interface.

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java (hunk 2 above)
Comment (robstryker) [reply bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java]: Fixing it

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java (hunk 3 above)
Comment (robstryker) [reply bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java]: Switched to BaseHTMLPrinter.  

Other options such as MickaelIsAwesomeHTMLPrinter and RobsLinguisticallyDeficientHTMLPrinter were rejected as too verbose.

**Comment by robstryker** [comment]
The test failures appear to be in the `Eclipse UI Tests ` bundle and do not appear related to these changes at all. But I welcome a second look.

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java (hunk 3 above)
Comment (mickaelistria) [reply bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java]: Who rejected those proposals? We should get the PMC to vote!
I was more wondering whether this class is actually a printer or something else. As a non-native english speaker, I'm naturally linguistically deficient in English, and to me, a printer is mostly 🖨️ and this class doesn't seem really related to 🖨️. I would go for a more typical Java name such as `HTMLBuilder` or as it's mostly stateless, something like `HTMLToolbox` or something like that. Do 👷 or 🧰 sound worse than 🖨️ for what we're trying to do here?

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java (hunk 1 above)
Comment (mickaelistria) [reply bundles/org.eclipse.text/src/org/eclipse/text/html/CoreRGB.java]: OK, I see that record cannot hide the canonical constructor (so we could still have checks with the record); which is a pity.

File: bundles/org.eclipse.text/src/org/eclipse/text/html/RGB.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+	/**
+	 * the green component of the RGB
+	 */
+	public int green;
+
+	/**
+	 * the blue component of the RGB
+	 */
+	public int blue;
```
Comment (mickaelistria) [inline bundles/org.eclipse.text/src/org/eclipse/text/html/RGB.java]: Can those 3 become final?

File: bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java (hunk 3 above)
Comment (laeubi) [reply bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java]: > I'm naturally linguistically deficient in English, and to me, a printer is mostly 🖨️ and this class doesn't seem really related to 🖨️

I think this is related to printing on the screen or similar like in "System.out.**print**ln" also java has **Print**Writer class...

Comment (mickaelistria) [reply bundles/org.eclipse.text/src/org/eclipse/text/html/CoreHTMLPrinter.java]: Yes, I get it. Yet I'm not sure this class is really about printing on the screen, it's more about building HTML strings as far as I could see.

**Comment by robstryker** [comment]
Changed to HTMLBuilder, made RGB fields final. 

Let's get this pushed plz ;)

**Pull request merged**